dask/dataframe/io/io.py

```diff
@@ -244,61 +244,75 @@ def from_pandas(
+    NotImplementedError
+        If passed either multi-indexed or null-indexed numeric data.
 
     See Also
     --------
     from_array : Construct a dask.DataFrame from an array that has record dtype
     read_csv : Construct a dask.DataFrame from a CSV file
     """
+    # Check data
+    if not has_parallel_type(data):
+        raise TypeError("Input must be a pandas DataFrame or Series.")
+
+    # Check index
     if isinstance(getattr(data, "index", None), pd.MultiIndex):
         raise NotImplementedError("Dask does not support MultiIndex Dataframes.")
 
-    if not has_parallel_type(data):
-        raise TypeError("Input must be a pandas DataFrame or Series.")
+    # Check data has no duplicate column names
+    # Even if this is eventually checked during construction, it's faster to check it at this point
+    if is_dataframe_like(data):
+        # Here it's assumed that a named index was originally a column, so we prevent a name clash
+        DataFrame._prevent_dupl_col_names(data, allow_index_name_in_cols=False)
 
+    # Check we have either chunksize or npartitions to know how to partition
     if (npartitions is None) == (chunksize is None):
         raise ValueError("Exactly one of npartitions and chunksize must be specified.")
-
-    nrows = len(data)
-
-    if chunksize is None:
-        if not isinstance(npartitions, int):
-            raise TypeError(
-                "Please provide npartitions as an int, or possibly as None if you specify chunksize."
-            )
-    elif not isinstance(chunksize, int):
+    if chunksize is None and not isinstance(npartitions, int):
         raise TypeError(
-            "Please provide chunksize as an int, or possibly as None if you specify npartitions."
+            "Please provide npartitions as an int, or set it to None if you specify chunksize."
+        )
+    if npartitions is None and not isinstance(chunksize, int):
+        raise TypeError(
+            "Please provide chunksize as an int, or set it to None if you specify npartitions."
         )
-
     name = name or ("from_pandas-" + tokenize(data, chunksize, npartitions))
 
-    if not nrows:
+    # If there are no rows, only 1 partition with unknown divisions
+    if (nrows := len(data)) == 0:
         return new_dd_object({(name, 0): data}, name, data, [None, None])
 
-    if data.index.isna().any() and not data.index.is_numeric():
-        raise NotImplementedError(
-            "Index in passed data is non-numeric and contains nulls, which Dask does not entirely support.\n"
-            "Consider passing `data.loc[~data.isna()]` instead."
-        )
-
+    # If we sort the index
     if sort:
+        # A non-numeric index should not have nulls, because sortedness is ambiguous in that case
+        if not data.index.is_numeric() and data.index.isna().any():
+            raise NotImplementedError(
+                "Index in passed data is non-numeric and contains nulls, which Dask does not entirely support.\n"
```

Comment:
Is there an issue connected to this? It seems to strong to me to simply disallow such data. Other sorting may also be ambiguous.

Reply:
Yes, having nulls in the index is problematic for several reasons: 1) partitions should be sorted, but it's ambiguous whether nulls should go before or after all non-null partitions, 2) every null should be in the same partition, but there are several "null-like" values, distinct from each other if they're compared (`None`, `nan`, `NaT`, etc.), 3) `None` is currently used to mean a division is unknown, and several inner workings of `fix_overlap` and related methods rely on either all divisions being `None`, on all divisions being non-`None`. This check was introduced by myself in MR #8963  cc @jsignell

Reply:
A `None` in the partitions list would not come from here, since these are values and those are counts. I don't think it's possible to have None, NaN and NaT in the same series, unless it is object type (non-string), in which case sorting will probably fail anyway. The ordering, in any case, should just be whatever pandas does.